### PR TITLE
chore(exp): Add pod autoscaler experiment used to check the scalability of the application pod

### PR DESCRIPTION
**Pull request by uditgaurav**
_Signed-off-by: Udit Gaurav <udit.gaurav@mayadata.io>_

## Issue:
   - fixes: https://github.com/litmuschaos/litmus/issues/1171

# Details:

- This PR is for adding a new experiment _"pod auto-scheduler"_ in litmus-go. The experiment aims to check the _ability of nodes_ to accommodate the number of replicas of the same application pod. We first scale the replicas of application pod then wait for a certain chaos duration to check if all pods comes to running state or not and this becomes the deciding factor for the experiment if it is passed or failed.

**It has two inputs for this :** 

  - **_Total Chaos Duration:_**  It is the time period upto which the experiment will wait and check for the replicas to be scheduled successfully in other words it is the timeout for this chaos.  
  - **_Replica Count:_** This is the number of replicas upto which we want to scale the application pod and test. 

```
For Example: 
If Total Chaos Duration is set to 600 sec and Replica count to 15 then the experiment will check till 10 mins 
to get all the replicas in Running state. 
```
### Node auto-scaling test: 

Now this experiment can be used for different scenarios as well, like for checking the Node auto-scaling feature. If the pods are unable to schedule then the nodes are getting auto-scaled or not.  

### pod auto-scaling check:

_pod auto-scaling:_ The _pod scaling check_ can be done by noting the amount of time taken by the pods to get scheduled after scaleup. The heavy application may takes larger chaos duration to get scheduled.

### Pre-Chaos and Post-Chaos Checks:

- The Pre-Chaos check for the experiment is to check the health of the application pods that is if all the replicas are in Running state. After the per chaos check the initial replica count is recorded. 
- Post Chaos-Check: It will again do a status check over the application pods this time after chaos. In both cases (Passed/Failed) we scale down the deployments to the original replicas after chaos injection.

### Abort:

In  case of anything goes wrong and you want to do [abort](https://docs.litmuschaos.io/docs/node-memory-hog/#abortrestart-the-chaos-experiment) chaos then this experiment supports that it means when we abort the experiment either by running abort command or deleting the experiment in middle, the experiment gets removed and the replica count scale down to original value(before chaos).

### Limitations:

- Currently it supports deployment type of application only. The support for statefulset will also be added very soon.

### CODE REVIEW

**Comment by saiyam1814** [comment]
This is a very good feature where we can test node and pod additions. 
I see lot of commits happen can you make it as a single commit? 
something like
git rebase -i HEAD~3 and follow the instructions in the interactive rebase

**Comment by saiyam1814** [comment]
So if I understand correct , it will check of a pod has insufficient node and then if it goes to running the test passes for node scaling?

**Comment by uditgaurav** [comment]
@saiyam1814  yes! it'll do that on account of two inputs, number of replicas and timeout (chaos duration).

**Comment by saiyam1814** [comment]
Great sounds good and I am sure community would really like this feature which is currently not there in any chaos tools

File: chaoslib/litmus/pod_autoscaler/pod-autoscaler.go [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+		return errors.Errorf("Unable to get the relicaCount of the application, err: %v", err)
+	}
+
+	//Waiting for the ramp time before chaos injection
+	if experimentsDetails.RampTime != 0 {
+		log.Infof("[Ramp]: Waiting for the %vs ramp time before injecting chaos", strconv.Itoa(experimentsDetails.RampTime))
+		waitForRampTime(experimentsDetails)
+	}
+	if err != nil {
+		return errors.Errorf("Unable to get the serviceAccountName, err: %v", err)
```
Comment (ispeakc0de) [inline chaoslib/litmus/pod_autoscaler/pod-autoscaler.go:38]: can we remove this?

File: chaoslib/litmus/pod_autoscaler/pod-autoscaler.go [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+	}
+
+	err = PodAutoscalerChaos(experimentsDetails, clients, replicaCount, appName, resultDetails, eventsDetails, chaosDetails)
+
+	if err != nil {
+		return errors.Errorf("Unable to perform autoscaling, due to %v", err)
+	}
+
+	err = AutoscalerReovery(experimentsDetails, clients, replicaCount, appName)
```
Comment (ispeakc0de) [inline chaoslib/litmus/pod_autoscaler/pod-autoscaler.go:47]: typo?

File: chaoslib/litmus/pod_autoscaler/pod-autoscaler.go [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+}
+
+//GetApplicationDetails is used to get the application name, replicas of the application
+func GetApplicationDetails(experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets) (string, int, error) {
+
+	var appReplica int
+	var appName string
+	applicationClient := clients.KubeClient.AppsV1().Deployments(experimentsDetails.ChaosNamespace)
+	// Get Deployment replica count
+	applicationList, err := applicationClient.List(metav1.ListOptions{})
```
Comment (ispeakc0de) [inline chaoslib/litmus/pod_autoscaler/pod-autoscaler.go:72]: Can we add len(applicationList) != 0 case?

File: chaoslib/litmus/pod_autoscaler/pod-autoscaler.go [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+func waitForRampTime(experimentsDetails *experimentTypes.ExperimentDetails) {
+	time.Sleep(time.Duration(experimentsDetails.RampTime) * time.Second)
+}
+
+//GetApplicationDetails is used to get the application name, replicas of the application
+func GetApplicationDetails(experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets) (string, int, error) {
+
+	var appReplica int
+	var appName string
+	applicationClient := clients.KubeClient.AppsV1().Deployments(experimentsDetails.ChaosNamespace)
```
Comment (ispeakc0de) [inline chaoslib/litmus/pod_autoscaler/pod-autoscaler.go:70]: Why are we using the client and then List in different lines is this intentional?

File: chaoslib/litmus/pod_autoscaler/pod-autoscaler.go [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+	ChaosStartTimeStamp := time.Now().Unix()
+	var endTime <-chan time.Time
+	timeDelay := time.Duration(experimentsDetails.ChaosDuration) * time.Second
+	failFlag := false
+	applicationClient := clients.KubeClient.AppsV1().Deployments(experimentsDetails.AppNS)
+	applicationDeploy, err := applicationClient.Get(appName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Errorf("Unable to get the application, err: %v", err)
+	}
+	for count := 0; count < int(experimentsDetails.ChaosDuration/2); count++ {
```
Comment (ispeakc0de) [inline chaoslib/litmus/pod_autoscaler/pod-autoscaler.go:130]: Any specific reason for iterate for `experimentsDetails.ChaosDuration/2` times?

can we use our `/pkg/utils/retry`?

File: chaoslib/litmus/pod_autoscaler/pod-autoscaler.go [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+				case <-endTime:
+					log.Infof("[Chaos]: Time is up for experiment: %v", experimentsDetails.ExperimentName)
+					endTime = nil
+					break loop
+
+				}
+
+			}
+
+			err = AutoscalerReovery(experimentsDetails, clients, replicaCount, appName)
```
Comment (ispeakc0de) [inline chaoslib/litmus/pod_autoscaler/pod-autoscaler.go:190]: typo

File: chaoslib/litmus/pod_autoscaler/pod-autoscaler.go [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+			if err != nil {
+				return errors.Errorf("Unable to perform autoscaling, due to %v", err)
+			}
+		} else {
+			break
+		}
+	}
+	if failFlag == true {
+		err = AutoscalerReovery(experimentsDetails, clients, replicaCount, appName)
```
Comment (ispeakc0de) [inline chaoslib/litmus/pod_autoscaler/pod-autoscaler.go:199]: typo

File: experiments/generic/pod-autoscaler/pod-autoscaler-k8s-job.yml [hunk 8]
```diff
@@ -0,0 +1,57 @@
+---
+apiVersion: batch/v1
+kind: Job
```
Comment (ispeakc0de) [inline experiments/generic/pod-autoscaler/pod-autoscaler-k8s-job.yml:3]: test.yaml & rbac.yaml?

File: chaoslib/litmus/pod_autoscaler/pod-autoscaler.go [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+	}
+	log.Info("[Info]: Application Started Scaling back")
+
+	applicationDeploy, err := clients.KubeClient.AppsV1().Deployments(experimentsDetails.AppNS).Get(appName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Errorf("Unable to get the application, err: %v", err)
+	}
+	failFlag := false
+	// Check for 30 retries with 2secs of delay
+	for count := 0; count < 30; count++ {
```
Comment (ispeakc0de) [inline chaoslib/litmus/pod_autoscaler/pod-autoscaler.go:241]: /pkg/utils/retry

File: chaoslib/litmus/pod_autoscaler/pod-autoscaler.go [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+				break
+			}
+
+			// signChan channel is used to transmit signal notifications.
+			signChan := make(chan os.Signal, 1)
+			// Catch and relay certain signal(s) to signChan channel.
+			signal.Notify(signChan, os.Interrupt, syscall.SIGTERM, syscall.SIGKILL)
+		loop:
+			for {
+				endTime = time.After(timeDelay)
```
Comment (ispeakc0de) [inline chaoslib/litmus/pod_autoscaler/pod-autoscaler.go:158]: Can we make this waiting time to the duration and remove the above steps of the outer loop?

**Comment by uditgaurav** [comment]
**Chaos Result** 

```
Name:         nginx-chaos-pod-autoscaler
Namespace:    default
Labels:       name=nginx-chaos-pod-autoscaler
Annotations:  <none>
API Version:  litmuschaos.io/v1alpha1
Kind:         ChaosResult
Metadata:
  Creation Timestamp:  2020-08-15T10:09:54Z
  Generation:          2
  Resource Version:    13007592
  Self Link:           /apis/litmuschaos.io/v1alpha1/namespaces/default/chaosresults/nginx-chaos-pod-autoscaler
  UID:                 ede02311-330e-41a8-8cf7-b29f38dc314f
Spec:
  Engine:      nginx-chaos
  Experiment:  pod-autoscaler
Status:
  Experimentstatus:
    Fail Step:                 N/A
    Phase:                     Completed
    Probe Success Percentage:  Awaited
    Verdict:                   Pass
Events:
  Type    Reason   Age    From                         Message
  ----    ------   ----   ----                         -------
  Normal  Summary  2m20s  pod-autoscaler-s6kmw6-q52t5  pod-autoscaler experiment has been Passed
```

**Chaos Engine**

File: chaoslib/litmus/pod_autoscaler/pod-autoscaler.go [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+	}
+	// Keeping a wait time of 10s after all pod comes in running state
+	// This is optional and used just for viewing the pod status
+	time.Sleep(10 * time.Second)
+
+	return nil
+}
+
+//AutoscalerRecovery scale back to initial number of replica
+func AutoscalerRecovery(experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, replicaCount int, appName string) error {
```
Comment (ksatchit) [inline chaoslib/litmus/pod_autoscaler/pod-autoscaler.go:160]: Deploy replica scale (up/down) can be a dedicated function.

File: chaoslib/litmus/pod_autoscaler/pod-autoscaler.go (hunk 9 above)
Comment (ksatchit) [reply chaoslib/litmus/pod_autoscaler/pod-autoscaler.go:189]: retry.go can be updated/used here

File: chaoslib/litmus/pod_autoscaler/pod-autoscaler.go [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+func PodAutoscalerChaos(experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, replicaCount int, appName string, resultDetails *types.ResultDetails, eventsDetails *types.EventDetails, chaosDetails *types.ChaosDetails) error {
+
+	applicationClient := clients.KubeClient.AppsV1().Deployments(experimentsDetails.AppNS)
+
+	replicas := int32(experimentsDetails.Replicas)
+	// Scale Application
+	retryErr := retries.RetryOnConflict(retries.DefaultRetry, func() error {
+		// Retrieve the latest version of Deployment before attempting update
+		// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+		result, err := applicationClient.Get(appName, metav1.GetOptions{})
```
Comment (ksatchit) [inline chaoslib/litmus/pod_autoscaler/pod-autoscaler.go:86]: `result` here indicates the appDeployment we have obtained via the client call. Can we name it as such (say, AppUnderTest) - result can be confusing considering there are other connotations to this  in our experiment.

File: chaoslib/litmus/pod_autoscaler/pod-autoscaler.go (hunk 11 above)
Comment (ksatchit) [reply chaoslib/litmus/pod_autoscaler/pod-autoscaler.go:160]: Tracked via issue #84

File: experiments/generic/pod-autoscaler/README.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
+## Experiment CR for the pod-autoscaler experiment
```
Comment (ksatchit) [inline experiments/generic/pod-autoscaler/README.md:1]: Lets keep the readme structure consistent with other experiments.

File: experiments/generic/pod-autoscaler/test/test.yml [hunk 14]
```diff
@@ -0,0 +1,54 @@
+apiVersion: batch/v1
+kind: Job
```
Comment (ksatchit) [inline experiments/generic/pod-autoscaler/test/test.yml:2]: The test.yaml is expected to be a deployment as per the latest scaffold utils (for use w/ okteto)

**Pull request merged**